staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 // Run spawns the secure http server. It only returns if stopCh is closed
 // or the secure port cannot be listened on initially.
 func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
-	delayedStopCh := make(chan struct{})
+	events := s.TerminationEvents
+
+	signalFn := s.overrideTerminationEventSignalFunc
```

Comment:
so, looking at the previous code https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go it seems like knowing when `delayedStopCh` and `drainedCh` were closed would allow to write unit tests and cover the entire `Run` method.

I'm just thinking if we could restore the previous version of the code and inject a channel factory instead.
The factory would allow for creating a new chan and wrapping already existing ones. 

Then we would change 
the `Run` method to accept the factory `Run(factoryCh chanFactory)`.
`make(chan struct{})` to `factoryChan.new("name")`
`close(...)` to `chanWrapperInstance.close()`
reading from a chan `<-` to calling the `wait` method



```
type chanFactory interface {
 new(name) chanWrapper
 wrap(chan) chanWrapper
}


type chanWrapper interface {
 wait() // blocking 
 close()
 rawChan()
}
```

After that we would get:
```

func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}, factoryCh chanFactory) error {
	delayedStopCh := factoryCh.new("delayedStopCh")
         stopCh = factoryCh.wrap(stopCh)

	go func() {
		deferdelayedStopCh.close()

		stopCh.wait()

		// As soon as shutdown is initiated, /readyz should start returning failure.
		// This gives the load balancer a window defined by ShutdownDelayDuration to detect that /readyz is red
		// and stop sending traffic to this server.
		close(s.readinessStopCh)

		time.Sleep(s.ShutdownDelayDuration)
	}()

	// close socket after delayed stopCh
	stoppedCh, err := s.NonBlockingRun(delayedStopCh.rawChan())
	if err != nil {
		return err
	}

         stoppedCh = factoryCh.wrap(stoppedCh)

	drainedCh := factoryCh.new("drainedCh")
	go func() {
		defer drainedCh.close()

		// wait for the delayed stopCh before closing the handler chain (it rejects everything after Wait has been called).
		delayedStopCh.wait()

		// Wait for all requests to finish, which are bounded by the RequestTimeout variable.
		s.HandlerChainWaitGroup.Wait()
	}()

	stopCh.wait()

	// run shutdown hooks directly. This includes deregistering from the kubernetes endpoint in case of kube-apiserver.
	err = s.RunPreShutdownHooks()
	if err != nil {
		return err
	}

	// Wait for all requests in flight to drain, bounded by the RequestTimeout variable.
	drainedCh.wait()
	// wait for stoppedCh that is closed when the graceful termination (server.Shutdown) is finished.
	stoppedCh.wait()

	return nil
}
```

Reply:
not sure how hard would it be to create these abstractions.

Reply:
A couple of notes:
- i have a followup PR where i have a filter that will need the termination event channel, so we can't create these channels inside the `Run` function, see here https://github.com/kubernetes/kubernetes/pull/101257/files#diff-207875acc7ee03f987aa626f53742c02323ed8f434fd40dfd929dffb23d3f1deR789.
```
	handler = genericfilters.WithHSTS(handler, c.HSTSDirectives)
	if c.KeepListeningDuringGracefulTermination {
		handler = genericfilters.WithNotAcceptingNewRequest(handler, c.TerminationEvents.WithName(NotAcceptingNewRequest).Wait())
	}
	handler = genericfilters.WithHTTPLogging(handler)
``

-

Reply:
Also, 
```
type chanWrapper interface {
 wait() // blocking 
 close()
 rawChan()
}
```

is similar to what we have in this PR
```
type TerminationSignal interface {
	Name() string
	Signal()
	Signaled() <-chan struct{}
}
```

- `wait` is similar to `Signaled` (except `Signaled` returns a `<- chan struct{}` as opposed to `wait` which blocks ) `Signaled` seems more idiomatic and more composable, 
- with `Signaled` we don't need `rawChan`
- `close` = `Signal`

If it makes it easier to read i can go with
```
type TerminationSignal interface {
	Name() string
	close()
	done() <-chan struct{}
}
```
Thoughts?